ddl/ddl_api.go

```diff
@@ -775,10 +784,38 @@ func buildColumnsAndConstraints(
 	return cols, constraints, nil
 }
 
+func getDefaultCollationForUTF8MB4(sessVars *variable.SessionVars, cs string) (string, error) {
+	if sessVars == nil || cs != charset.CharsetUTF8MB4 {
+		return "", nil
+	}
+	defaultCollation, err := sessVars.GetSessionOrGlobalSystemVar(context.Background(), variable.DefaultCollationForUTF8MB4)
+	if err != nil {
+		return "", err
+	}
+	return defaultCollation, nil
+}
+
+// GetDefaultCollation returns the default collation for charset and handle the default collation for UTF8MB4.
+func GetDefaultCollation(sessVars *variable.SessionVars, cs string) (string, error) {
```

Comment:
Why is this function in `ddl` package, but not in `charset` package?

I suggest to modify the current `charset.GetDefaultCollation`, add an argument `sessVars`. We have only 10+ reference to the `charset.GetDefaultCollation`, and most of them have `sctx` in the environment, so I think it won't be a big change to do this refractor. But I'm not sure whether it's the expected behavior... I see this PR doesn't touch the `expression` package, but `(*convertFunctionClass).getFunction` will also call `GetDefaultCollation`. Is it expected?

Besides the `GetDefaultCollation`, I also found `GetDefaultCollationLegacy` used in parser. I'm not sure whether it should also be modified :thinking: .

Reply:
I put it here because it is currently only used for DDL operations, and the PR description explains the scope of this variable's influence, that is, it should not affect the expression.

Reply:
> it should not affect the expression

Well this actually affects the expression, we should expect the following result:

```
tidb> set @@default_collation_for_utf8mb4='utf8mb4_0900_ai_ci';
Query OK, 0 rows affected, 1 warning (0.00 sec)

tidb> select _utf8mb4'a' = _utf8mb4'A', _utf8mb4'a' = _utf8mb4'a  ';
+---------------------------+-----------------------------+
| _utf8mb4'a' = _utf8mb4'A' | _utf8mb4'a' = _utf8mb4'a  ' |
+---------------------------+-----------------------------+
|                         1 |                           0 |
+---------------------------+-----------------------------+
1 row in set (0.01 sec)
```

As mentioned in the PR desc:

> Any statement containing a string literal of the form _utf8mb4'some text' without a COLLATE clause.

So I think @YangKeao 's comment is reasonable.

Reply:
> it should not affect the expression

Well this actually affects the expression, we should expect the following result:

```
tidb> set @@default_collation_for_utf8mb4='utf8mb4_0900_ai_ci';
Query OK, 0 rows affected, 1 warning (0.00 sec)

tidb> select _utf8mb4'a' = _utf8mb4'A', _utf8mb4'a' = _utf8mb4'a  ';
+---------------------------+-----------------------------+
| _utf8mb4'a' = _utf8mb4'A' | _utf8mb4'a' = _utf8mb4'a  ' |
+---------------------------+-----------------------------+
|                         1 |                           0 |
+---------------------------+-----------------------------+
1 row in set (0.01 sec)
```

As mentioned in the PR desc:

> Any statement containing a string literal of the form _utf8mb4'some text' without a COLLATE clause.

So I think @YangKeao 's comment is reasonable.

Reply:
Yes, as I said above, currently only DDL is affected, and PR also describes that "This PR doesn't support item 4."

Reply:
Sorry I missed the description!